scripts/exp/extract-deps.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as fs from 'fs';
+import * as path from 'path';
+import * as yargs from 'yargs';
+import {
+  ExportData,
+  extractDependenciesAndSize,
+  extractFunctionsAndClasses
+} from './extract-deps.helpers';
+
+// Small command line app that builds a JSON file with a list of the
+// dependencies for each individual exported type. This can then be used to
+// verify dependencies in tree-shakeable SDKs.
+
+const argv = yargs.options({
+  types: {
+    type: 'string',
+    demandOption: true,
+    desc: 'The location of the index.d.ts file that describes the Public API'
+  },
+  bundle: {
+    type: 'string',
+    demandOption: true,
+    desc:
+      'The location of transpiled JavaScript bundle that contains all code for the SDK'
+  },
+  output: {
+    type: 'string',
+    demandOption: true,
+    desc: 'The location to write the JSON output to'
+  }
+}).argv;
+
+async function buildJson(publicApi: string[], jsFile: string): Promise<string> {
+  const result: { [key: string]: ExportData } = {};
+  for (const exp of publicApi) {
+    result[exp] = await extractDependenciesAndSize(exp, jsFile);
+  }
+  return JSON.stringify(result, null, 4);
+}
+
+const publicApi = extractFunctionsAndClasses(path.resolve(argv.types));
```

Comment:
It doesn't handle reexports, e.g. `export * from 'src/impl'`. We can generate a rollup d.ts file to sidestep the problem, or we need to traverse the AST to get all exports.

Reply:
You are extracting APIs from the d.ts file, not the source file, right? Currently we don't bundle d.ts files into a single file, so there can be reexports from other d.ts files.

As we plan to have multiple entry points in firestore next, it will be easier to not have a single bundled source file so it's easier to share code between different entry points. Why would transitive dependencies be a problem though?

Reply:
The public APIs are extracted from a handcrafted index.d.ts. Are we moving away from hand-crafted types? Note that this is not meant to work for all of Firebase, but only for each specific SDK.

FWIW, as more SDKs adopt this, we probably have to greatly expand the feature set of this tooling. I wonder if it makes sense to wait until there is a need, which can then happen together with CI integration?

Reply:
I leave that decision to teams. But generally we don't need to handcraft d.ts anymore in the modular world, because we control what gets exported in the source code, and we can generate d.ts files with only public APIs from it automatically. 

I generate index.d.ts from source code and use it for `@firebase/app-exp`. It allows me to keep documentation in the source code as well. I expect most of our SDKs to use automatically generated d.ts files in the modular world.

Sure. that sounds reasonable to me.

Reply:
It would be interesting to see if we can hide SDK-public internals from the public typings without splitting up all Firestore classes into abstract Public-API-only base classes and implementation classes. I will keep an eye on this as we rewrite the client.